Zend/zend_execute.c

```diff
@@ -1364,13 +1340,19 @@ static zend_never_inline zend_long zend_check_string_offset(zval *dim, int type
 	if (UNEXPECTED(Z_TYPE_P(dim) != IS_LONG)) {
 		switch(Z_TYPE_P(dim)) {
 			case IS_STRING:
-				if (IS_LONG == is_numeric_string(Z_STRVAL_P(dim), Z_STRLEN_P(dim), NULL, NULL, -1)) {
-					break;
-				}
-				if (type != BP_VAR_UNSET) {
-					zend_error(E_WARNING, "Illegal string offset \"%s\"", Z_STRVAL_P(dim));
+			{
+				/* allow errors in string offset for BC reasons */
+				if (IS_LONG == is_numeric_string(Z_STRVAL_P(dim), Z_STRLEN_P(dim), &offset, NULL, true)) {
+					/* emit Illegal string warning on leading numerical string */
+					if (0 == is_numeric_string(Z_STRVAL_P(dim), Z_STRLEN_P(dim), NULL, NULL, false)
```

Comment:
That's indeed a better approach.

Sadly yeah I kept a sort of -1 mode to make the BC less harsh so that it has a better chance of passing.
I'm basically relegating the promotion of these warnings (which were notices) to TypeError to PHP 9